### DISCUSSION TITLE
chapter_appendix/contribution/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#7] [Announcements]
# chapter_appendix/contribution/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_appendix/contribution/

**Comment by HopeLightning** [comment]
HiHiHi，偶然发现了这个很有意思的项目，顺着看完的“写在前面”的内容，感觉在阅读流程上可以优化一下。

### 优化内容

- 需求：网页端的目录结构更扁平一点，鼠标悬停展开目录的方式更舒服
    - 原因
        - 目前的层级结构不太直观会产生误会；（比如我这次看完了“写在前面”内容，点开目录发现已经到最后了，一下子会有种难道只更新到这里的困惑感）
    - 具体参考
        - 目前markdown的列表的展示和大纲笔记的折叠方式；

P.S. 另外有没有考虑增加上下篇的跳转按钮呢？不过这个感觉优先级不太高，如果目录功能能优化就已经很方便了

**Reply by krahets** [reply]
谢谢您的建议！目前竖屏下的目录确实是这样，我看看有没有办法解决。如果是横屏的话，目录是默认展开的～

**Reply by krahets** [reply]
以及关于上下篇跳转按钮，这个后续会加上～

**Reply by seven1240** [reply]
看了，非常好。我本地跑了一下，请教两个问题：
1）文档里的图片和表格 id 是怎么算同来的，本地跑的 Docker 里没有 id 关联，而在线版的都将“上图”、“下图”之类的内容算出来了，显示“如图1-2”之类的。
2）PDF是怎么编译出来的？使用什么命令？

**Comment by xiaobinqt** [comment]
大神，你这书用什么主题写的 ?

**Reply by krahets** [reply]
哈喽，是 mkdocs

**Comment by WSL0809** [comment]
哇，原来是 K 神

**Comment by Logenleedev** [comment]
想问一下这个图是用什么软件做的啊～ 感觉看着好舒服啊

**Comment by FangYuan33** [comment]
Hello, K神你好，我是方圆，我想问一下本书有没有规划中待写的小节，我想在假期的时候帮忙写某个小节，我平时自己也会写博客儿，文字表达应该是可以的，期待你的回复

**Reply by krahets** [reply]
哈喽，感谢哇，方便加我微信聊嘛 krahets-jyd

**Reply by amberzhang39** [reply]
请问我可以加您微信吗？非常感谢写了这个本书，很有帮助！

**Reply by krahets** [reply]
@yingz030906 当然可以啊哈哈

**Comment by WSL0809** [comment]
哈喽，我发现书里的代码示例没有rust版本，但是仓库里有rust版本的代码示例，请问有没有考虑把rust的代码也加到书中呢？

**Reply by krahets** [reply]
Hi，Rust 代码仍在开发中，等待内容比较完整 + 测试后就会放到文档里～ 可以在 #15 看下还有哪些章节需要补充～

**Reply by WSL0809** [reply]
okok，感谢回复

**Comment by bananapomelo** [comment]
可以考虑出C版本的PDF嘛  工作不能接触网络  比较遗憾没法工作中学

**Reply by krahets** [reply]
很想，但是目前 c 代码还不完整，得靠大家努努力补完代码

**Reply by seven1240** [reply]
现在完整了吗？或者这个不完整有不完整列表不？或者从哪里找到，我可以贡献一把。

**Comment by zhuasdfg** [comment]
学完打卡，DP和贪心真的是太难了

**Comment by WhiteRhinos** [comment]
感谢

**Comment by Wokening** [comment]
学完打卡，感谢大佬！

**Comment by xwshiba** [comment]
谢谢作者们！！发现整本书除了数据结构那章，并没有提到位操作相关的内容，请问会有意向写作或加上吗？

**Reply by krahets** [reply]
位运算等有时间写上！

**Comment by staytacit** [comment]
作者你好，感谢奉献了这么精彩的教程；本学员从力扣.cn上收益良多，一图胜千言。
如果有意参与创作，请问教程里的精美图片是用什么工具、工作流创作出来的？
我查了issue、readme前言等地方没找到说明，望解答，谢谢

**Reply by krahets** [reply]
ppt keynote 这类软件都可以画的

**Reply by staytacit** [reply]
好的 谢谢

**Comment by cocowise** [comment]
开源图书有个问题是，由于参与者众多，可能无意中参杂入一些完全复制过来的段落。例如我看到第四章末尾，有个Q&A，要是复制上这句“创建一个新的整数对象时，CPython 首先会检查它是否在−5 和 256 之间。如果是，CPython 就会直接返回现有的整数对象，而不必浪费时间创建新的对象。”虽然能很好回答问题，但这复制自a copyrighted book

**Reply by krahets** [reply]
谢谢！本项目的每个 PR 都会严格 review ，避免出现类似情况。

**Comment by anyeyongzhou** [comment]
学完打卡，对算法有一个最基本的认识了，下面就是去不断刷题，锻炼自己的思维

**Comment by KailynnnWuuu** [comment]
请问可以加一个 做笔记的 功能吗  阅读时可以批注、加高亮、下划线，然后有个收藏夹集中放批注高亮过的笔记，这样方便回顾重点

**Reply by krahets** [reply]
Hi，这是一个很好的功能，可惜暂时做不了。你可以尝试用浏览器插件实现～

**Reply by KailynnnWuuu** [reply]
好的 感谢 我之前使用的是 Weava Highlighter - PDF & Web 这个插件，现在不好用了，总是卡顿，给浏览网页带了了不好的体验，所以想试试别的插件了

**Comment by jerrygzy** [comment]
大概看了一下内容, 提几个建议:

1. 增加nim语言的版本
2. 增加算法应用的篇幅, 尤其是操作系统,数据库, 密码学, 编译器 相关的篇幅
3. 增加算法在软件工程的实践, 尤其是与AI对话的提示词

最后推荐一个不错的资源: dmitrysoshnikov.teachable.com

**Reply by krahets** [reply]
谢谢建议！新增语言只能等待 PR 贡献了～

**Comment by drhetianhao** [comment]
学习中，如果有推荐的例题来练习就更好啦

**Reply by krahets** [reply]
建议先把书一次性刷完～想刷题的话可以看看 https://github.com/krahets/LeetCode-Book

**Comment by plwater** [comment]
按照给的指令部署到本机 Docker 后没有正常访问，网页显示为空。
截图如下，希望能够解决一下，感谢作者的无私工作。
![ykAdtiF](https://i.imgur.com/ykAdtiF.jpg)
![U4CKbei](https://i.imgur.com/U4CKbei.jpg)
![jMuJWTe](https://i.imgur.com/jMuJWTe.jpg)

**Reply by krahets** [reply]
谢谢反馈，已于 https://github.com/krahets/hello-algo/commit/774c27c47ed88fca717955fad936ed9a0a53dcf7 修复

**Reply by plwater** [reply]
重新部署后访问成功，感谢修复🌹

**Comment by kaibo-li** [comment]
大佬您好，我本地docker部署之后，没有可视化运行，需要怎么弄

**Reply by plwater** [reply]
因为你没有提供更多信息，我的建议是『删除 Docker 容器和镜像、拉取最新代码、重新按照作者给定指令执行、注意端口号不要冲突、以及其他的细节』，如果还是不行，建议检查 Docker 是否正常，再多做几遍上面的步骤，肯定会解决你的问题。

**Comment by kaibo-li** [comment]
您好，重新尝试了几次，安装也没报错，本地访问就只有可视化运行没有，其他一切正常
 
> 因为你没有提供更多信息，我的建议是『删除 Docker 容器和镜像、拉取最新代码、重新按照作者给定指令执行、注意端口号不要冲突、以及其他的细节』，如果还是不行，建议检查 Docker 是否正常，再多做几遍上面的步骤，肯定会解决你的问题。

**Comment by githubxzmclj** [comment]
一直部署失败，求个大佬帮忙看一下
Service 'hello-algo' failed to build: error pulling image configuration: Get "https://production.cloudflare.docker.com/registry-v2/docker/registry/v2/blobs/sha256/dd/dd04d4f70e7c3fc372ddf5a78d703edcd80d8f51a8fe35f7c7ca64840fae3747/data?verify=1721414638-zf%2FRfKUviGSALE4JPkgDlYuxQDQ%3D": dial tcp 174.37.175.229:443: connect: connection refused

**Reply by githubxzmclj** [reply]
好吧，成功了，多弄了几个镜像源就好了

**Comment by cloudsmithy** [comment]
请问这个前端是自己写的嘛，还是用了什么样的插件？

**Comment by xiangyang0831** [comment]
问一下，为什么我部署的页面中，长一些的代码就会是这样的

[file]{array_hash_map}-[class]{array_hash_map}-[func]{}

在源码的markdown文件中对应的地方是
```src
[file]{array_hash_map}-[class]{array_hash_map}-[func]{}
```
这样的，不知道为何不显示对应的源码的代码 而只是显示
[file]{array_hash_map}-[class]{array_hash_map}-[func]{}

**Reply by krahets** [reply]
可以看一下 markdown 语法，如果要显示 code block，可以用 ``` ``` 来包裹

**Comment by loveRose** [comment]
算法时间复杂度定义应该是有误，准确说应该是算法主体执行步骤量，说是算法执行消耗的时间这个肯定是有误的，虽然网上博客很多都使用时间这个标准，但是应该谨慎的说不能用时间标准来衡量算法的时间复杂度，比如O(1)的时间复杂度，能说它是执行消耗多少时间吗？这只是说执行任意大小的数据，所执行的步骤都是固定的，如果是指时间，就算相同的算法连续执行多次都可能会消耗不同的时间。

**Reply by krahets** [reply]
Hi，书中哪里说 算法的时间复杂度的定义是 “算法执行消耗的时间”？

**Comment by zhiwenhu471** [comment]
2026.1.10 第一次看完，开始刷题

**Comment by ly3638** [comment]
up可以增添一下红黑树和b树吗

**Reply by krahets** [reply]
谢谢建议！这两部分内容有些难，感觉有些偏离本书的定位了，所以一直没下定决心写。在已有内容中，感觉 AVL 树已经触碰到难度的 upperbound 了
